mod-source-record-manager-server/src/main/java/org/folio/services/flowcontrol/RawRecordsFlowControlServiceImpl.java

```diff
@@ -85,80 +84,99 @@ public void trackChunkReceivedEvent(String tenantId, Integer recordsCount) {
-      Collection<KafkaConsumerWrapper<String, String>> rawRecordsReadConsumers = consumersStorage.getConsumersByEvent(DI_RAW_RECORDS_CHUNK_READ.value());
 
-      rawRecordsReadConsumers.forEach(consumer -> {
-        if (consumer.demand() == 0) {
-          consumer.resume();
+    LOGGER.info("resumeIfThresholdAllows :: Tenant: [{}]. Current state:{}, History state: {}",
+      tenantId, currentState.get(tenantId), historyState.get(tenantId));
 
-          LOGGER.info("resumeIfThresholdAllows:: Tenant: [{}]. Kafka consumer - id: {}, subscription - {} is resumed for all tenants, because {} met threshold {}",
-            tenantId, consumer.getId(), DI_RAW_RECORDS_CHUNK_READ.value(), this.currentState, recordsThreshold);
+    historyState.putIfAbsent(tenantId, Pair.of(0,0));
+    if (historyState.get(tenantId).getKey().equals(currentState.get(tenantId))) {
+      Integer previousValue = historyState.get(tenantId).getValue();
+      historyState.put(tenantId, Pair.of(historyState.get(tenantId).getKey(), ++previousValue));
+      LOGGER.debug("resumeIfThresholdAllows :: Tenant: [{}]. Increase counter. Current state:{}, History state: {}",
+        tenantId, currentState.get(tenantId), historyState.get(tenantId));
+    } else {
+      historyState.put(tenantId, Pair.of(currentState.get(tenantId),0));
+    }
+
+    consumersStorage.getConsumersByEvent(DI_RAW_RECORDS_CHUNK_READ.value())
+      .forEach(consumer -> {
+        LOGGER.info("resumeIfThresholdAllows :: Tenant: [{}]. DI_RAW_RECORDS_CHUNK_READ. " +
+            "ConsumerId:{}, Demand:{}, Current state:{}, History state: {}",
+          tenantId, consumer.getId(), consumer.demand(), currentState.get(tenantId), historyState.get(tenantId));
+        if (((consumer.demand() == 0) && (currentState.get(tenantId) < recordsThreshold)) ||
+          (historyState.get(tenantId).getKey() > 0 && historyState.get(tenantId).getValue() > 0)) {
+          if (consumer.demand() > maxSimultaneousChunks) {
+            consumer.pause(); //set demand to 0, because fetch can only add new demand value when demand > 0
+          }
+          consumer.fetch(maxSimultaneousChunks);
+
+          LOGGER.info("resumeIfThresholdAllows :: Tenant: [{}]. Fetch: DI_RAW_RECORDS_CHUNK_READ. " +
+            "ConsumerId:{}, Demand:{}, Current state:{}, History state:{}", tenantId, consumer.getId(), consumer.demand(),
+            currentState.get(tenantId), historyState.get(tenantId));
+
+          //counters reset
+          historyState.put(tenantId, Pair.of(0, 0));
+          currentState.put(tenantId, 0);
         }
       });
+  }
+
+  private void initFetchMode(String tenantId) {
+
+    // A read stream is either in "flowing" or "fetch" mode
+    // initially the stream is in "flowing" mode
+    // when the stream is in "flowing" mode, elements are delivered to the handler
+    // when the stream is in "fetch" mode, only the number of requested elements will be delivered to the handler
```

Comment:
on the handler is done processing a fetch, will another fetch be performed automatically?

Reply:
Not. Fetch could not be started automatically. For this, I use a scheduler for checking a consumer and when a consumer is on pause (demand == 0), the scheduler will increase demand.

Reply:
Reached out to @Aliaksandr-Fedasiuk on Slack. The scheduler is not responsible for calling fetch on the consumer each time. It is responsible for waking a paused consumer. The consumer will continue to fetch messages on a cadence.